### PR TITLE
[zkp] Test infinity throughout our test suite as well

### PR DESCRIPTION
**Pull request by huitseeker**
As it turns out, Arkworks also has a non-standard representation of the infinity point. This integrates the infinity point in all randomly-generated tests, and corrects the discrepancies.

### CODE REVIEW

**Review by kchalkias — COMMENTED** [review]
Seems good code-wise, but need to study the details of the fixed values before approving: 0x40, mask the flag bits, and 10% maybe infinity?

File: fastcrypto-zkp/src/conversions.rs [hunk 1]
```diff
@@ -143,8 +143,18 @@ pub fn bls_g1_affine_to_blst_g1_affine(pt: &BlsG1Affine) -> blst_p1_affine {
...
-    let mut g1 = blst_p1_affine::default();
+    // BLST disagrees with Arkworks on the uncompressed representation of the infinity point on G1
+    // https://github.com/supranational/blst/blob/6382d67c72119d563975892ed49ba32e92d3d0da/src/e1.c#L153-L162
+    // the infinity bit notwithstanding,
+    // Arkworks: x = Fq::zero(), y = Fq::one()
+    // BLST: x = y = Fq::zero()
+    // BLST follows the standard here (see https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-10#appendix-C)
+    if pt.infinity {
+        tmp2[0] = 0x40; // set the infinity bit
+        tmp2[48..].copy_from_slice(&[0u8; 48]); // zeroize the rest of the point
```
Comment (kchalkias) [inline fastcrypto-zkp/src/conversions.rs:154]: Ok now I get it, but I guess for converting y=Fq::one() ->  Fq::zero() you only needed to set one bit Vs copy the whole 48 bytes slice?

File: fastcrypto-zkp/src/conversions.rs [hunk 2]
```diff
@@ -154,12 +164,25 @@ pub fn blst_g1_affine_to_bls_g1_affine(pt: &blst_p1_affine) -> BlsG1Affine {
...
+    // BLST: x = y = Fq::zero()
+    // BLST follows the standard here (see https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-10#appendix-C)
+    let y = if infinity {
+        Fq::one()
+    } else {
+        Fp384::from_be_bytes_mod_order(&out[48..])
+    };
+    let x = {
+        // Mask away the flag bits (which Arkworks doesn't use)
+        out[0] &= 0b0001_1111;
```
Comment (kchalkias) [inline fastcrypto-zkp/src/conversions.rs:182]: I read the RFC, thanks, it makes sense. Because the curve uses 381 bits, the first 3 bits in a 384bit stream (48bytes) are used as flags, but Arkworks didn't take that into account and instead they used the 0,1 point as infinity :(

File: fastcrypto-zkp/src/conversions.rs [hunk 3]
```diff
@@ -176,6 +199,16 @@ pub fn bls_g2_affine_to_blst_g2_affine(pt: &BlsG2Affine) -> blst_p2_affine {
...
     };
+    // BLST disagrees with Arkworks on the uncompressed representation of the infinity point on G2
+    // https://github.com/supranational/blst/blob/6382d67c72119d563975892ed49ba32e92d3d0da/src/e2.c#L194-L203
+    // the infinity bit notwithstanding,
+    // Arkworks: x = Fq2::zero(), y = Fq2::one()
+    // BLST: x = y = Fq2::zero()
+    // BLST follows the standard here (see https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-10#appendix-C)
+    if pt.infinity {
+        tmp2[0] = 0x40; // set the infinity bit
+        tmp2[96..].copy_from_slice(&[0u8; 96]); // zeroize the rest of the point
```
Comment (kchalkias) [inline fastcrypto-zkp/src/conversions.rs:210]: ditto, I think you don't need to copy the full slice but just set one bit to convert `Fq2::one()` -> `Fq2::zero()` (all zero bytes)

File: fastcrypto-zkp/src/conversions.rs [hunk 4]
```diff
@@ -532,38 +579,65 @@ pub(crate) mod tests {
...
+
     pub(crate) fn arb_bls_g1_affine() -> impl Strategy<Value = BlsG1Affine> {
         // slow, but good enough for tests
-        arb_bls_fr().prop_map(|s| {
-            BlsG1Affine::prime_subgroup_generator()
-                .mul(s.into_repr())
-                .into_affine()
+        (arb_bls_fr(), any::<f32>()).prop_map(|(s, maybe_infinity)| {
+            if maybe_infinity < 0.1 {
+                // 10% chance of being the point at infinity
```
Comment (kchalkias) [inline fastcrypto-zkp/src/conversions.rs:592]: Can you please explain or provide a reference for this 10% probability?

**Comment by kchalkias** [comment]
> As it turns out, Arkworks also has a non-standard representation of the infinity point. This integrates the infinity point in all randomly-generated tests, and corrects the discrepancies.

I remember mentioning this edge case in a previous PR, thanks!

File: fastcrypto-zkp/src/conversions.rs (hunk 4 above)
Comment (huitseeker) [reply fastcrypto-zkp/src/conversions.rs]: Good catch, this is a clunky and erroneous attempt at making this a minority case. The reason it kinda-sorta worked experimentally is that proptest divides floats in 7 strategies: https://docs.rs/proptest/1.0.0/proptest/num/f32/constant.ANY.html
I've refactored this (and other instances) to actually reflect what I intended to do.

File: fastcrypto-zkp/src/conversions.rs (hunk 1 above)
Comment (huitseeker) [reply fastcrypto-zkp/src/conversions.rs:154]: That works, if I'm being given a legitimate Arkworks point produced by the Arkworks library. 

Except Arkworks conveys the infinity using a separate flag here, so by using custom code, or simply abusing the somewhat permissive deserialization, the caller could send whatever they want in the second coordinate.

File: fastcrypto-zkp/src/conversions.rs (hunk 3 above)
Comment (huitseeker) [reply fastcrypto-zkp/src/conversions.rs:210]: Same as [the above](https://github.com/MystenLabs/fastcrypto/pull/149#discussion_r1003412930): I think it's safer to not assume the caller will use Arkworks conventions, and return the correct result over the whole input range.

File: fastcrypto-zkp/src/conversions.rs (hunk 1 above)
Comment (kchalkias) [reply fastcrypto-zkp/src/conversions.rs:154]: interesting, so it's malleable, pretty bad IMO

Comment (kchalkias) [reply fastcrypto-zkp/src/conversions.rs:154]: I'm happy to err on the safe side, so your approach meets the bar.

File: fastcrypto-zkp/src/conversions.rs (hunk 3 above)
Comment (kchalkias) [reply fastcrypto-zkp/src/conversions.rs:210]: another option is to set this a const and just use that. IMO that would safer vs having code to reconstruct the infinity point every single time.

Comment (huitseeker) [reply fastcrypto-zkp/src/conversions.rs:210]: Right, so given the `blst` definitions all call into FFI functions (all non-const), I don't have the ability to make a const definition, here. But I can certainly make auxiliary functions for clarity.

File: fastcrypto-zkp/src/conversions.rs [hunk 5]
```diff
@@ -154,16 +176,48 @@ pub fn blst_g1_affine_to_bls_g1_affine(pt: &blst_p1_affine) -> BlsG1Affine {
...
         blst_p1_affine_serialize(out.as_mut_ptr(), pt);
     }
-    let infinity = out[0] & (1 << 6) != 0;
-    BlsG1Affine::new(
-        Fp384::from_be_bytes_mod_order(&out[..48]),
-        Fp384::from_be_bytes_mod_order(&out[48..]),
-        infinity,
-    )
+    let infinity = out[0] == 0x40;
+    println!("out: {:?}, infinity: {:?}", out, infinity);
```
Comment (kchalkias) [inline fastcrypto-zkp/src/conversions.rs:180]: forgotten `println`

**Review by mystenadmin — APPROVED** [review]
LGTM

**Review by kchalkias — APPROVED** [review]
